### PR TITLE
Only merge valid entitlements keys from provisioning profiles

### LINKED ISSUE #798: Store submission fails with iCloud entitlements
When a provisioning profile has iCloud entitlements enabled, the buck signed binary will not pass store submission. A typical iCloud provisioning profile will have entitlements similar to:

```lang=xml
<key>Entitlements</key>
<dict>
    <key>keychain-access-groups</key>
    <array>
        <string>4ZX4Z3MVHG.*</string>
    </array>
    <key>application-identifier</key>
    <string>4ZX4Z3MVHG.com.appleseedinc.MyProject</string>
    <key>com.apple.developer.ubiquity-kvstore-identifier</key>
    <string>4ZX4Z3MVHG.*</string>
    <key>com.apple.developer.icloud-services</key>
    <string>*</string>
    <key>com.apple.developer.icloud-container-environment</key>
    <array>
        <string>Development</string>
        <string>Production</string>
    </array>
    <key>com.apple.developer.icloud-container-identifiers</key>
    <array>
        <string>iCloud.com.appleseedinc.MyProject</string>
        <string>iCloud.com.appleseedinc.container1</string>
    </array>
    <key>com.apple.developer.icloud-container-development-container-identifiers</key>
    <array>
        <string>iCloud.com.appleseedinc.MyProject</string>
        <string>iCloud.com.appleseedinc.container1</string>
    </array>
    <key>com.apple.developer.ubiquity-container-identifiers</key>
    <array>
        <string>iCloud.com.appleseedinc.MyProject</string>
        <string>iCloud.com.appleseedinc.container1</string>
    </array>
    <key>com.apple.developer.team-identifier</key>
    <string>4ZX4Z3MVHG</string>
    <key>aps-environment</key>
    <string>production</string>
</dict>
```

The default entitlements handling in buck will combine the specified entitlements file with the entitlements in the provisioning profile. This will result in both the `com.apple.developer.icloud-container-development-container-identifiers` and `com.apple.developer.ubiquity-kvstore-identifier` keys being added to the signing entitlements. Trying to submit a binary with either of these present will result in the following iTunes store errors:

```
ERROR ITMS-90045: "Invalid Code Signing Entitlements. Your application bundle's signature contains code signing entitlements that are not supported on iOS. Specifically, key com.apple.developer.icloud-container-development-container-identifiers in Payload/Blah.app is not supported."
```

and:

```
ERROR ITMS-90211: Invalid Code Signing Entitlements. The signature for your app bundle contains entitlement values that are not supported. For the com.apple.developer.ubiquity-kvstore-identifier entitlement, the value must start with the prefix provided by Apple in the provisioning profile, followed by characters that are uppercase or lowercase Roman letters [A-Z, a-z], the digits 0 through 9, dot, or hyphen, and not contain any wildcard characters. Specifically, value 4ZX4Z3MVHG.* for the key com.apple.developer.ubiquity-kvstore-identifier in Payload/Blah.app is not supported.
```
 
I'm not sure where the behaviour of merging the entitlements came from, is this mirroring what Xcode would do? It seems more correct to not merge the entitlements, would this result in a different error though, do we always need to merge certain keys from the provisioning profile?

### PR DESCRIPTION
**Pull request by rmaz**
Fixes #798. Use a blacklist of keys that should not be merged from provisioning profile entitlements into xcent files.

### CODE REVIEW

**Comment by facebook-github-bot-7** [comment]
By analyzing the blame information on this pull request, we identified @ryu2, @sdwilsh and @k21 to be potential reviewers.

**Comment by ryu2** [comment]
@facebook-github-bot shipit

File: test/com/facebook/buck/apple/ProvisioningProfileMetadataTest.java [hunk 1]
```diff
@@ -98,4 +100,18 @@ public void testSplitAppID() throws Exception {
...
+  public void testFilteredEntitlementsStripOut() throws Exception {
+    assumeTrue(Platform.detect() == Platform.MACOS);
+    Path testdataDir = TestDataHelper.getTestDataDirectory(this).resolve("provisioning_profiles");
+    Path testFile = testdataDir.resolve("sample.mobileprovision");
+
+    ProvisioningProfileMetadata data =
+        ProvisioningProfileMetadata.fromProvisioningProfilePath(testFile);
+
+    assertTrue(data.getEntitlements().containsKey("com.apple.security.application-groups"));
+    assertFalse(data.getMergeableEntitlements().containsKey("com.apple.security.application-groups"));
```
Comment (ryu2) [inline test/com/facebook/buck/apple/ProvisioningProfileMetadataTest.java]: Make sure the line is <= 100 chars.

File: test/com/facebook/buck/apple/ProvisioningProfileCopyStepTest.java [hunk 2]
```diff
@@ -168,4 +175,58 @@ public void shouldSetProvisioningProfileFutureWhenStepIsRun() throws Exception {
...
+    );
+    step.execute(executionContext);
+
+    ProvisioningProfileMetadata selectedProfile = step.getSelectedProvisioningProfileFuture().get();
+    ImmutableMap<String, NSObject> profileEntitlements = selectedProfile.getEntitlements();
+    assertTrue(profileEntitlements.containsKey( "com.apple.security.application-groups"));
+
+    Optional<String> xcentContents = projectFilesystem.readFileIfItExists(xcentFile);
+    assertTrue(xcentContents.isPresent());
+    NSDictionary xcentPlist = (NSDictionary)PropertyListParser.parse(xcentContents.get().getBytes());
```
Comment (ryu2) [inline test/com/facebook/buck/apple/ProvisioningProfileCopyStepTest.java]: space after `(NSDictionary)` cast

File: test/com/facebook/buck/apple/ProvisioningProfileCopyStepTest.java [hunk 3]
```diff
@@ -168,4 +175,58 @@ public void shouldSetProvisioningProfileFutureWhenStepIsRun() throws Exception {
...
+        ProvisioningProfileStore.fromSearchPath(testdataDir),
+        outputFile,
+        xcentFile,
+        codeSignIdentityStore
+    );
+    step.execute(executionContext);
+
+    ProvisioningProfileMetadata selectedProfile = step.getSelectedProvisioningProfileFuture().get();
+    ImmutableMap<String, NSObject> profileEntitlements = selectedProfile.getEntitlements();
+    assertTrue(profileEntitlements.containsKey( "com.apple.security.application-groups"));
```
Comment (ryu2) [inline test/com/facebook/buck/apple/ProvisioningProfileCopyStepTest.java]: nit: remove space before `"com`

File: test/com/facebook/buck/apple/ProvisioningProfileCopyStepTest.java [hunk 4]
```diff
@@ -168,4 +175,58 @@ public void shouldSetProvisioningProfileFutureWhenStepIsRun() throws Exception {
...
```
Comment (ryu2) [inline test/com/facebook/buck/apple/ProvisioningProfileCopyStepTest.java]: nit: remove space before `"com`

**Comment by ryu2** [comment]
Thanks!  Everything looks good, just a few style nits.  Make sure no line is > 100 chars; otherwise our linter complains.

**Comment by facebook-github-bot-0** [comment]
@rmaz updated the pull request.

**Comment by facebook-github-bot-0** [comment]
@rmaz updated the pull request.

**Pull request closed without merging**